src/apify_client/_utils.py

```diff
@@ -0,0 +1,101 @@
+import random
+import time
+from datetime import datetime, timezone
+from typing import Any, Callable, Dict, Tuple, TypeVar, cast
+
+from ._errors import ApifyApiError
+
+PARSE_DATE_FIELDS_MAX_DEPTH = 3
+PARSE_DATE_FIELDS_KEY_SUFFIX = 'At'
+
+NOT_FOUND_STATUS_CODE = 404
+NOT_FOUND_TYPE = 'record-not-found'
+NOT_FOUND_ON_S3 = '<Code>NoSuchKey</Code>'
+
+
+def _to_safe_id(id: str) -> str:
+    return id.replace('/', '~')
+
+
+def _parse_date_fields(data: Dict) -> Dict:
+    return cast(Dict, _parse_date_fields_internal(data))
+
+
+def _parse_date_fields_internal(data: object, max_depth: int = PARSE_DATE_FIELDS_MAX_DEPTH) -> object:
+    if max_depth < 0:
+        return data
+
+    if isinstance(data, list):
+        return list(map(lambda item: _parse_date_fields_internal(item, max_depth - 1), data))
+
+    if isinstance(data, dict):
+        def parse(item: Tuple[str, object]) -> Tuple[str, object]:
+            (key, value) = item
+            parsed_value = value
+            if key.endswith(PARSE_DATE_FIELDS_KEY_SUFFIX) and isinstance(value, str):
+                try:
+                    parsed_value = datetime.strptime(value, '%Y-%m-%dT%H:%M:%S.%fZ').replace(tzinfo=timezone.utc)
+                except ValueError:
+                    pass
+            else:
+                parsed_value = _parse_date_fields_internal(value, max_depth - 1)
+            return (key, parsed_value)
+
+        return dict(map(parse, data.items()))
+
+    return data
+
+
+def _pluck_data(parsed_response: Any) -> Dict:
+    if isinstance(parsed_response, dict) and 'data' in parsed_response:
+        return cast(Dict, parsed_response['data'])
+
+    raise ValueError('The "data" property is missing in the response.')
+
+
+T = TypeVar('T')
+BailType = Callable[[Exception], None]
+
+
+def _retry_with_exp_backoff(
```

Comment:
What about turning this in decorator? I think it would make it less verbose and easier to use.

Decorator example (for iterating function N times, I chose this as it's not that long and illustrates the point).
```
from typing import Any, Callable
from functools import wraps


def iterate_decorator(num_of_iterations: int) -> Callable:
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def iterate_fn(*args, **kwargs) -> Any: 
            for attempt in range(num_of_iterations):
                print(f"Iteration {attempt} in decorator")
                func(*args, **kwargs)
        return iterate_fn
    return decorator


@iterate_decorator(num_of_iterations = 3)
def print_foo_decorated(arg: str, kw_arg: str):
    print('foo decorated')


print_foo_decorated('arg', kw_arg='kw_arg')
```

When executed, this returns
```
Iteration 0 in decorator
foo decorated
Iteration 1 in decorator
foo decorated
Iteration 2 in decorator
foo decorated
```

Reply:
Also, I'd not force decorated function  to have `bail` and `attempt` arguments. I think they should be able to have any arguments. Instead, I think we should pass to decorate itself on which exceptions it should continue and on which it should simply throw. Let's abstract that from the function decorated.